File: python/repl/monkey/environment.py
class Environment(object):
    store = {}
    outer = None
    def __init__(self):
        self.store = {}
        self.outer = None

    def Get(self, name):
        check = name in self.store
        if not check and self.outer:
            return self.outer.Get(name)
        if check:
            return self.store[name], check
        else:
            return None, check

    def Set(self, name, val):
        check = name in self.store
        self.store[name] = val
        return val, check

def NewEnvironment():
    return Environment()

def NewEnclosedEnvironment(outer):
    env = NewEnvironment()
    env.outer = outer;
    return env

File: cpp/vm/advm/environment.cpp
#include "environment.h"

Environment* newEnvironment() {
    Environment* env = new Environment();
    return env;
}

Environment* newEnvironmentWithOuter(Environment* outer) {
    Environment* env = new Environment();
    env->outer = outer;
    return env;
}

AdObject* Environment::get(std::string key) {
    if (store.find(key) != store.end()) {
        return store.at(key);
    }
    if (outer->store.find(key) != outer->store.end()) {
        return outer->store.at(key);
    }
    return NULL;
}

void Environment::save(std::string key, AdObject* obj) {
    
}